openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py

```diff
@@ -0,0 +1,210 @@
+
+    Using this approach makes the connection between the ObjectTag
+    and system defined taxonomy as hardcoded and can't be changed.
+    """
+
+    system_defined_taxonomy_id = None
+
+    def _check_system_taxonomy(self, taxonomy: Taxonomy = None):
+        """
+        Validates if the taxonomy is system-defined and match
+        with the name stored in the object tag
+        """
+        return (
+            bool(taxonomy) and
+            taxonomy.system_defined and
+            taxonomy.id == self.system_defined_taxonomy_id
+        )
+
+
+class OpenSystemObjectTag(OpenObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Free-text object tag used on system-defined taxonomies
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+    
+class ClosedSystemObjectTag(ClosedObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Object tags linked to a closed system-taxonomy
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+
+class ModelObjectTag(OpenSystemObjectTag):
+    """
+    Object tags used with tags that relate to the id of a model
+    
+    This object tag class is not registered as it needs to have an associated model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = None
+
+
+    def _check_taxonomy(self):
+        """
+        Validates if has an associated Django model that has an Id
+        """
+        if not super()._check_taxonomy():
+            return False
+    
+        if not self.tag_class_model:
+            return False
+
+        # Verify that is a Django model
+        if not issubclass(self.tag_class_model, models.Model):
+            return False
+
+        # Verify that the model has 'id' field
+        try:
+            self.tag_class_model._meta.get_field('id')
+        except FieldDoesNotExist:
+            return False
+
+        return True
+    
+    def _check_instance(self):
+        """
+        Validates if the instance exists
+        """
+        try:
+            intance_id = int(self.value)
+        except ValueError:
+            return False
+        return self.tag_class_model.objects.filter(id=intance_id).exists()
+
+    def _check_tag(self):
+        """
+        Validates if the instance exists
+        """
+        if not super()._check_tag():
+            return False
+        
+        # Validates if the instance exists
+        if not self._check_instance():
+            return False
+        
+        return True
+
+
+class UserObjectTag(ModelObjectTag):
+    """
+    Object tags used on taxonomies associated with user model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = get_user_model()
```

Comment:
I really like how you used `external_id` to store the language code for the language object tag , and then used the full language name as the value. So when we actually fetch an ObjectTag for display, we have a nice human-readable value to show.

I'd like to support something like that here with model tags. So that when we have a tag that references a User or Organization, we can show the User/Organization full name instead of the ID. Plus, that name will always be up-to-date, because we're fetching it from the source table.

So I'm wondering, can we do something like this:

```python
class ObjectTag:
    @property
    def value_pretty(self):
        """
        Returns a human-friendly version of the tag value stored here.
        
        Subclasses should override this if they don't store human-readable values.
        """
        return self.value
```

And then we can have here:
```
class ModelObjectTag(...):
    @property
    def tag_class_model_pretty_field(self):
        """
        Name of the field to use on the tag_class_model instance to which stores the "pretty value" of the tag.
        """
        raise NotImplementedError

    @property
    def value_pretty(self):
        """
        Returns a human-friendly version of the tag value stored here.
        
        Subclasses should override this if they don't store human-readable values.
        """
        value = self.value
        try:
            instance = self.tag_class_model.objects.get(pk=value)
            value = getattr(instance, self.tag_class_model_pretty_field, value)
        except self.tag_class_model.DoesNotExist:
            # Means the tag is invalid, but that's ok here
            pass
            
        return value
...

class UserObjectTag(ModelObjectTag):
    @property
    def tag_class_model_pretty_field(self):
        """
        Name of the field to use on the tag_class_model instance to which stores the "pretty value" of the tag.
        """
        return "full_name"
```

Reply:
Updated to use this approach https://github.com/open-craft/openedx-learning/pull/3#discussion_r1267622247